### PR TITLE
Breakout history

### PR DESCRIPTION
**Pull request by nhunzaker**
This PR extracts the history API into a separate module. Ideally, this could eventually be overridden. It also lays the groundwork for a consistent set of lifecycle methods for transactions.

### CODE REVIEW

File: src/histories/Branches.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * Branches History
+ * Stores transaction history as a tree.
+ */
+
+import Tree from '../Tree'
+
+let Branches = function (options={}) {
+  this.base = {}
+  this.tree = new Tree()
+  this.maxHistory = options.maxHistory || -Infinity
```
Comment (averyvery) [inline src/histories/Branches.js]: why the -?

**Comment by averyvery** [comment]
:+1: really sharp

File: src/histories/Branches.js (hunk 1 above)
Comment (nhunzaker) [inline src/histories/Branches.js]: By default, I never want to keep any history. This feels more permanent than `0`.

Also I never miss a chance to use Infinity :)

**Comment by nhunzaker** [comment]
Giving my own :+1:  to circumvent our slackbot :P

**Pull request merged**